src/expr/src/interpret.rs

```diff
@@ -0,0 +1,1184 @@
+    /// when our values set is empty or     contains a single datum, but when it contains a range,
+    /// we need a little metadata about the function to be able to infer a useful result.
+    fn flat_map(
+        self,
+        is_monotonic: Monotonic,
+        mut datum_map: impl FnMut(Datum<'a>) -> ResultSpec<'a>,
+    ) -> ResultSpec<'a> {
+        let null_spec = if self.nullable {
+            datum_map(Datum::Null)
+        } else {
+            ResultSpec::nothing()
+        };
+
+        let error_spec = ResultSpec {
+            fallible: self.fallible,
+            ..Self::nothing()
+        };
+
+        let values_spec = match self.values {
+            Values::Empty => ResultSpec::nothing(),
+            // If this range contains a single datum, just call the function.
+            Values::Within(min, max) if min == max => datum_map(min),
+            // If this is a range of booleans, we know all the values... just try them.
+            Values::Within(Datum::False, Datum::True) => {
+                datum_map(Datum::False).union(datum_map(Datum::True))
+            }
+            // Otherwise, if our function is monotonic, we can try mapping the input
+            // range to an output range.
+            Values::Within(min, max) => match is_monotonic {
+                Monotonic::Yes => {
+                    let min_column = datum_map(min);
+                    let max_column = datum_map(max);
+                    if min_column.nullable
+                        || min_column.fallible
+                        || max_column.nullable
+                        || max_column.fallible
+                    {
+                        ResultSpec::anything()
+                    } else {
+                        min_column.union(max_column)
+                    }
+                }
+                Monotonic::Maybe | Monotonic::No => ResultSpec::anything(),
+            },
+            // TODO: we could return a narrower result for eg. `Values::Nested` with all-`Within` fields.
+            Values::Nested(_) | Values::All => ResultSpec::anything(),
+        };
+
+        null_spec.union(error_spec).union(values_spec)
+    }
+}
+
+/// [Abstract interpretation](https://en.wikipedia.org/wiki/Abstract_interpretation) for
+/// [MirScalarExpr].
+///
+/// [MirScalarExpr::eval] implements a "concrete interpreter" for the expression type: given an
+/// expression and specific column values as input, it returns a specific value for the output.
+/// This could be reimplemented using this trait... but it's most useful for "abstract"
+/// interpretations of the expression, where we generalize about sets of possible inputs and outputs.
+/// See [Trace] and [ColumnSpecs] for how this can be useful in practice.
+pub trait Interpreter {
+    type Summary: Clone + Debug + Sized;
+
+    /// A column of the input row.
+    fn eval_column(&self, id: usize) -> Self::Summary;
```

Comment:
nit: just my personal preference here, but I would drop the `eval_` prefix here in line with the "object algebra" encoding from literature (see the interfaces defind in [Extensibility for the Mases](https://www.cs.utexas.edu/~wcook/Drafts/2012/ecoop2012.pdf) and [Typed Tagless Final Interpreters](https://okmij.org/ftp/tagless-final/course/lecture.pdf)).

Reply:
The prefix is slightly historical, but I think the best argument for keeping it is that we may want a second flavour of the interpreter methods, like @mgree alludes to [here](https://github.com/MaterializeInc/materialize/pull/18932#discussion_r1180629891).

That's easy enough to change again later on, though... so I'll make the change unless @mgree prefers otherwise?

Reply:
I'm easy! I have a similar naming scheme in the typechecker: [Typecheck::typecheck()](https://github.com/MaterializeInc/materialize/blob/3e5d1996b6b859bd70b48b354027a27febd97e95/src/transform/src/typecheck.rs#L475) works on relation expressions, while [Typecheck::typecheck_scalar()](https://github.com/MaterializeInc/materialize/blob/3e5d1996b6b859bd70b48b354027a27febd97e95/src/transform/src/typecheck.rs#L1038) works on scalars and [Typecheck::typecheck_aggregate()](https://github.com/MaterializeInc/materialize/blob/3e5d1996b6b859bd70b48b354027a27febd97e95/src/transform/src/typecheck.rs#L1128) works on aggregates.

Reply:
Cool! Pushed up a commit with the rename.